db/compaction/compaction_picker_level.cc

```diff
@@ -572,11 +574,49 @@ bool LevelCompactionBuilder::PickFileToCompact() {
   const std::vector<FileMetaData*>& level_files =
       vstorage_->LevelFiles(start_level_);
 
+  uint64_t start_lvl_bytes_no_compacting = 0;
+  uint64_t curr_bytes_to_compact = 0;
+  uint64_t start_lvl_max_bytes_to_compact = 0;
+
+  if (ioptions_.compaction_pri == kRoundRobin && start_level_ != 0) {
+    // For round-robin compaction priority, we have 3 constraints when picking
+    // multiple files.
+    // Constraint 1: We can only pick consecutive files
+    // 	-> Constraint 1a: When a file is being compacted (or some input files
+    //                    are being compacted after expanding, we cannot
+    // choose it and have to stop choosing more files
+    //  -> Constraint 1b: When we reach the last file (with largest keys), we
+    //                    cannot choose more files (the next file will be the
+    //                    first one)
+    // Constraint 2: We should ensure the total compaction bytes (including the
+    //               overlapped files from the next level) is no more than
+    //               mutable_cf_options_.max_compaction_bytes
+    // Constraint 3: We try our best to pick as many files as possible so that
+    //               the post-compaction level size is less than
+    //               MaxBytesForLevel(start_level_)
```

Comment:
Should there be a Constraint 4 related to picking minimal files to split into the target number of subcompactions? A TODO is fine with me since Constraint 2 probably suffices (coincidentally, IMO) for most use cases today.

The reason for that is in case a level is hugely oversized, we don't really want to pick all the files necessary to make it not oversized, even if `max_compaction_bytes` allows it. That's because  it would dedicate too much compaction resources to fixing one level at a time, while other oversized levels stay the same or grow. Then the size ratio between consecutive levels would be unbalanced, and probably swing back and forth when compactions finish.

It is true that Constraint 2 (`max_compaction_bytes`) can prevent it, but so far that option is generally good to increase as long as space is available for larger compactions. Without a Constraint 4, users would additionally need to think about not setting it too high for compaction efficiency.

Reply:
Sure. I will add a TODO there.